### DISCUSSION TITLE
Suggestion: Add ComponentToString and ComponentToBytes

### DISCUSSION THREAD

**Discussion by JosueAD95** [a-h/templ#472] [Ideas]
Hello,

I was working in a personal project that use gorilla Websockets. I decided to update the project to use templ, I was using the function [WriteMessage](https://github.com/gorilla/websocket/blob/v1.5.1/conn.go#L773) that receive a slice of bytes. I started searching a way to get a []bytes for a Component and the closes I got was the function [ToGoHTML](https://github.com/a-h/templ/blob/v0.2.543/runtime.go#L775), I knew I can convert template.HTML to a string and them to []bytes. After that, I saw the code of the function and release that from the buffer I can get []bytes without getting a template.HTML first and I implemented the function in my Project.

I thought I could be helpful to have a function that converts a templ.Component to a string and other one that do the same just to a []bytes. I would like created a PR with this two functions, basically those will be similar to the ToGoHTML implementation. Please let me know what do you think.

Kind regards

**Comment by TAR5** [comment]
Yes, it would be really useful if the templ.Component implemented the .String() method.

My usecase would be this:
`<div data-my-attr={ myComponent.String() }></div>`
(I want to use the rendered component as an attribute value)

Afaik there is no out-of-the-box way to do this.

**Reply by joerdav** [reply]
I don't think we would add `String() string` to `templ.Component` as that would cause other peoples custom components to no longer be valid, I also think it is powerful for the interface to be a single function interface.

For now I would recommend writing a `ToString` function of your own, rather than it being a build in function.

```go
func TemplString(t templ.Component) (string, error) {
    var b bytes.Buffer
    if err := t.Render(context.Background(), &b); err != nil {
        return "", err
    }
    return b.String(), nil
}
```

`<div data-my-attr={ TemplString(myComponent) }></div>`
